src/storage/lock_manager/deadlock.rs

```diff
@@ -31,71 +31,102 @@ const TXN_DETECT_INFO_TTL: u64 = 120000;
 
 #[derive(Default)]
 struct DetectTable {
-    wait_for_map: HashMap<u64, Vec<Lock>>,
+    // txn_ts => (lock_ts, Vec<lock_hash>)
+    wait_for_map: HashMap<u64, HashMap<u64, Vec<u64>>>,
 }
 
 impl DetectTable {
-    /// Return deadlock key hash if deadlocked
+    /// Returns the deadlock key hash which causes deadlock.
     pub fn detect(&mut self, txn_ts: u64, lock_ts: u64, lock_hash: u64) -> Option<u64> {
         let _timer = DETECTOR_HISTOGRAM_VEC.detect.start_coarse_timer();
         TASK_COUNTER_VEC.detect.inc();
 
-        if let Some(deadlock_key_hash) = self.do_detect(txn_ts, lock_ts) {
+        // If txn_ts is waiting for lock_ts, it won't cause deadlock.
+        if self.register_if_existed(txn_ts, lock_ts, lock_hash) {
+            return None;
+        }
+
+        // Memorize the searching path.
+        let mut visited: HashSet<u64> = HashSet::default();
```

Comment:
Good idea

Reply:
So could we change it in the future?

Reply:
It's hard to replace it with a `RefCell` field. And I have benched it, it can't gain an obvious improvement.

Reply:
This doesn't improve too much since it may just save a few allocations. So it doesn't matter if we don't do it. By the way I don't think  `RefCell` is needed here...

Reply:
`do_detect` must use the immutable reference of `self`. If `visited` is a field, it must be 
 a mutable reference. I can only think of this solution.🤦🏻‍♂️